### ISSUE TITLE
[Proposal] Scheduler improvements for mass-hosting Laravel sites

### ISSUE THREAD

**Issue by uberbrady** [laravel/ideas#908]
I work with Grokability, who develops [Snipe-IT](/snipe/snipe-it), a Laravel-backed asset-management software platform. Laravel is amazing, and I don't think Snipe-IT would've been able to be developed without it. Thanks a million for developing it! That being said, we have encountered an edge-case with the Scheduler that we wanted to ask about.

We have several tasks we want to execute for our hosted customers using Laravel Scheduler, but have not been able to. (Open-source users of our app seem to be able to use Laravel Scheduler just fine, without any issues.)

First off, when we configure Scheduler via Cron as specified in the Laravel documentation, our servers begin performing poorly. A server has nearly 100 Laravel apps on it, and each one at exactly zero seconds past the minute tries to execute a PHP script - the machine basically grinds itself into the ground (sudden spike in memory usage, all other tasks slow down, system begins to eat into swap, it gets even slower - by the time the next minute comes up, the tasks launch again - and it gets worse.) Needless to say, we don't do that anymore.

So the second thing we tried to do was to set each Laravel instance to fire off `scheduler:run` on a different minute of the hour. That succeeded at not setting the server on fire, but the actual scheduled tasks do not ever execute. I suppose what's happening is that the various tasks are scheduled to run "once per day" which probably translates to exactly-midnight, or something similar.

One workaround that someone might consider would be to just use Cron to directly schedule the tasks - and this is, in fact, what we do. But it's not ideal; what's great about using Laravel Scheduler is that we can make code changes which then make schedule changes, or add or remove tasks, accordingly. Right now, if we decide to add, remove, or change a new scheduled task, we will have to add/remove/change the corresponding cronjob(s) for the associated Laravel instance. It'd be much easier and lower-maintenance for us to simply have the tasks working within Laravel Scheduler as specified.

The solution that I like for this is permitting Laravel users to run the scheduler at any frequency they desire - with `* * * * *` being the default, as it is today. But permitting them to run the scheduler less often, and having the scheduler try and "catch up" when it sees that it should've run something, but hasn't yet. Of course, if you have a Scheduler task that needs to run every minute, then running the Scheduler only once per hour wouldn't make any sense, so that would have to be made clear in the documentation. If whatever implementation that would handle this would require too much state-management, we would be fine with it being made optional and activated via `.env` vars or whatever.

If this is something that there would be interest in merging, we could certainly try to work on an implementation. Hopefully, we could get some pointers on where to look, first, of course :) Also, we would hope that this could be merged into a 5.4 Laravel version, as that's the version we're working with right now.

**Comment by sisve** [comment]
This can currently not be provided as a package since the dependency on the Schedule class is hard-coded in Kernel::defineConsoleSchedule().

https://github.com/laravel/framework/blob/1bc8dd90218d382e516425e7f4abb28fafb33dc1/src/Illuminate/Foundation/Console/Kernel.php#L95-L102

The `schedule:run` artisan command just calls `$schedule->dueEvents(...)` so no changes seems to be needed there.

https://github.com/laravel/framework/blob/1bc8dd90218d382e516425e7f4abb28fafb33dc1/src/Illuminate/Console/Scheduling/ScheduleRunCommand.php#L52

So "all" you have to do is ...

1. Ceate a CustomSchedule class that derives from the original.
2. Override your Kernel::defineCodeSchedule() and use the CustomSchedule class.
3. Implement an awesome CustomSchedule::dueEvents(...) method that keeps state somewhere so it knows when it was last called, and uses this to implement the magic.

**Comment by uberbrady** [comment]
Those are some great pointers, and that's a great way to do it!

We're in sort of a weird situation, in that we have to serve two masters - our open-source users, and our customers (they pay for hosting). We never want to add stuff that screws up our open-source users, or add stuff for the open-source people that our customers can't use.

So having a custom scheduler probably wouldn't mess up our open-source users too badly, but I'd also be curious to hear if this is an issue that other Laravel users have. If it's just us, we could certainly build out something just as you describe and it would probably be fine for both our open-source users as well as our hosted customers. But if there might be others who would want a similar type of solution, I'd love to see it added to the core - maybe as something you need to enable with `.env`, or maybe just 'on by default'. And schedule-tracking could probably just be done with simple files in the `./storage` directory, I suspect.

So - is this something that would possibly benefit other Laravel users? I'd imagine that launching a script every minute would affect more than just us.

**Comment by laurencei** [comment]
I dont understand this.

If you dont fire the Scheduler every minute - then you'll miss tasks. But you then say

> But permitting them to run the scheduler less often, and having the scheduler try and "catch up" when it sees that it should've run something, but hasn't yet.

So in the end you still run the same number of jobs. Your not really spreading the load much differently. Either way - if your clients schedule commands - then need to run.

**Comment by mattstauffer** [comment]
This is a really interesting challenge. I’m curious as well to see whether other people have this issue. 

My gut says no because it’s not a really common scenario for Laravel developers. But who knows! People working this way might be hiding out there. 

With regard to how to fix the solution, I like @sisve’s solution. You are queuing up the jobs that happen in the in between moments so you can use the cron scheduling to offset the load.

@laurencei while that may be true (get a bigger server) in many contexts, it’s not always the case. I’m confident Grokability know what they’re doing for their specific context. The bigger question here is, is this a unique need for them or a more common need? Like I said, my gut says it’s unique. But my gut isn’t always right.

**Comment by laurencei** [comment]
Sorry - I didnt mean it to sound harsh. What I meant was "delaying" scheduling doesnt solve the overall workload problem.

And delaying the scheduling is a dangerous logical rabbit hole.

**Example:**
If a customer/command schedules a command to run `* * * * *` - then they have an expecation it will run once a minute.  If you hold it off, and "catch up" - then you might run the command 5 times in one minute at a later time. Or you run it "once" later one. Either way - it's a change to the logical expectation of the scheduler - and it'll cause all sorts of issues because it will be unexpected behaviour.

**Comment by laurencei** [comment]
A slightly better approach would be to use the `scheduler->queue()` function - and have all 100 applications simulatenously push their jobs onto the queue. The server *should* be able to cope with that.

Then you can scale your queue worker infrastructure if required - and monitor it with Horizon or similar.

**Comment by laurencei** [comment]
Lastly - @uberbrady - do you schedule 100 crons? i.e. are you running

```
* * * * * php /app1/artisan schedule:run >> /dev/null 2>&1
* * * * * php /app2/artisan schedule:run >> /dev/null 2>&1
* * * * * php /app3/artisan schedule:run >> /dev/null 2>&1
```

If so - one other way to *slightly* offset the load is have each cron sleep slightly before running:


```
* * * * * sleep 1; php /app1/artisan schedule:run >> /dev/null 2>&1
* * * * * sleep 2; php /app2/artisan schedule:run >> /dev/null 2>&1
* * * * * sleep 3; php /app3/artisan schedule:run >> /dev/null 2>&1
```
As discussed here: https://stackoverflow.com/a/4420811/1317935

That will give each app a second to kick off before the next one. Note - dont sleep for more than around 50s in total - otherwise by the time the app tries to run it'll be too late and you'll get other sorts of issues.

**Comment by sisve** [comment]
Dispatching jobs would be weird since he would need either 100 queue workers, one for every customer, or a shared queuing system that captured and forwarded the customer context to the jobs, things like configurations.

A weird and slightly irrelevant thought:

I know that Jenkins supports crontab expressions with a hash (probably based on the project name), something like "0 0 * * *" would execute at midnight everyday, while "H H * * *" would execute once per day, but it's not clear when. Maybe there's a crontab system for your operating system that has similar support?

I imagine that Laravel could implement something like that for cases like "H/5 * * * *" (every five minutes, but spread out), but it wouldn't really work for the every-minute things. The original problem seems related to spreading out jobs within the same minute, not over several minutes...

**Comment by laurencei** [comment]
> Dispatching jobs would be weird since he would need either 100 queue workers, one for every customer

I guess that depends if they already have queues in place for the customers. If they do - then its an easy solution.

**Comment by uberbrady** [comment]
My problem is just wanting to have 'generic scheduler support' so that I have software control of scheduler tasks without having to manage 100 crontabs. I have no intention of trying to run a 'real' task every minute; the few tasks I have now run daily. I still want the ability to have hourly tasks, eventually, though. I'm fine with saying "well, if you only run your scheduler once an hour, you can only do hourly (or greater) tasks." My catch-up comment was about the fact that Hourly is probably implemented as `$x:00` and so if my scheduler artisan command doesn't run at exactly 0 minutes past the hour, it won't run an hourly task, ever.

We tried to have a 100 crons (on an early, pre-production server). Let's just say "that didn't go well" and leave it at that.

And, of course - if I'm scheduling some task to run every single minute, and then I try to batch them up to try and catch up once an hour, that would be terrible. I have no intention of doing that. And maybe, even, Laravel Scheduler could yell at you if it saw that kind of thing - e.g. it knows the last time it ran, it knows that it's running this time, it sees that's equal to 60 minutes, and it sees that you have a per-minute task scheduled. So it yells at you and does nothing.

The rule could be "If you want to run tasks once a minute, you have to run scheduler once a minute. If you want to run tasks hourly, you have to run scheduler at least once an hour." and so on.

And, yes, I could just run scheduler once an hour on the `:00` of each hour, for every single Laravel instance, with n-seconds in between  - but I suspect that's also a ticking time bomb - if my daily process fires off, it will fire off for every instance, and if it takes a while - kerplow. I've blown up my server again.

In answer to another question - no, we don't create a queuing system for each customer nor do we have a centralized queue. Honestly, I don't see a need for one, for us, yet? I'm not averse to such a thing in the future if we were to need it. I wouldn't want to saddle our open-source customers (who are running single-tenant) with such a requirement though.

Hopefully that clears up what my use case is here. I'd be happy to answer any more questions on that, as well. And I would definitely love to hear if other people have similar problems.

**Comment by sisve** [comment]
> I wouldn't want to saddle our open-source customers (who are running single-tenant) with such a requirement though.

The queueing solution/workaround could easily be implemented in a way where your single-tenant setups uses the `sync` queue driver for the scheduled jobs. This would execute the jobs synchronously (read: from the scheduler callback) instead of dispatching them to a queue.

**Comment by TvL2386** [comment]
Hey guys,

I wanted to have my own scheduler because I'm dockerizing my laravel app. I don't have cron.
So I made `php artisan schedule:daemon` https://gist.github.com/TvL2386/67fe243079cbdb6c3e7d0c0a11509782

Check the code. It will execute `php artisan schedule:run` on the 0th second every minute. Works great for me. 
I spent an hour to have it call it internally without having to invoke a child process.
I got to:
```
use Illuminate\Console\Scheduling\Schedule;
use Illuminate\Console\Scheduling\ScheduleRunCommand;
$schedule = app(Schedule::class);
(new ScheduleRunCommand($schedule))->handle();
```
instead of `php artisan schedule:run`, but that throws an exception:
```
PHP Error:  Call to a member function isDownForMaintenance() on null in /myapp/vendor/laravel/framework/src/Illuminate/Console/Scheduling/Event.php on line 265
```

Anyway: If you could get that to work, you would NOT have to boot the application every minute for every customer.

And I have already implemented what was said above: My scheduler creates Jobs. So the scheduler itself is not executing stuff that takes a lot of time, basically because I'm worried I would miss the next minute mark and skip a minute. I'm scheduling cron('*/15 * * * *') for example and don't want to miss an invocation.

Kind regards!

**Comment by taylorotwell** [comment]
Probably no plans to bring this into the framework.
